zPages proposal (#110)
open-telemetry/oteps: text/0110-z-pages.md

# zPages: general direction (#110)

Make zPages a standard OpenTelemetry component.

## Motivation

ZPages are in-process web pages that display collected data from the process they are attached to. They are used to provide in-process diagnostics without the need of any backend to examine traces or metrics. Various implementations of zPages are widely used in many environments. The standard extensible implementation of zPages in OpenTelemetry will benefit everybody.

## Explanation

> This OTEP is a request to get a general approval for zPages development as an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). See [opencensus.io/zpages](https://opencensus.io/zpages/) for the overview of zPages.

Reviewer: `opentelemetry-specification#62` or `opentelemetry-specification#632`?

Author: thank you!

## Internal details

> This is a request for a general direction approval, not specific details.

Reviewer: I think this is insufficient detail to request an "approval". I would prefer to establish goals (dos and donts) and requirements. Also a discussion of pros and cons would be appropriate for an OTEP, considering that there is prior art and presumably prior learnings.

Author: I have added principles. Does it sound OK now? I really want to simplify the process of experimentation lightweight. As you pointed - there is a prior art and hopefully it will guide the experiment to the right direction.

# zPages: general direction (#110)

Make zPages a standard OpenTelemetry component.

## Motivation

> ZPages or self-introspection debug pages are in-process web pages that display collected data from the process they are attached to. They are used to provide in-process diagnostics without the need of any backend to examine traces or metrics. Various implementations of zPages are widely used in many environments. The standard extensible implementation of zPages in OpenTelemetry will benefit everybody.

Reviewer 2: Is this `zPages` or `ZPages`?

Author: I think `zPages`. Just upper cased it in the beginning of a sentence. I changed the beginning of a sentence

## Explanation

This OTEP is a request to get a general approval for zPages development as an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). See [opencensus.io/zpages](https://opencensus.io/zpages/) for the overview of zPages.

## Internal details

This is a request for a general direction approval. There are a few principles for the development:

> 1. ZPages framework MUST NOT be hardcoded into OpenTelemetry SDK.

Reviewer 2: It is not clear whether zPages is a "component", "exporter" or "framework".

Author: clarified. added a sentence 

> Implementation of zPages includes multiple components - data collection (sampling, filtering, configuration), storage and aggregation, and a framework to expose this data.

and rephrased bullet points

# zPages: general direction (#110)

Make zPages a standard OpenTelemetry component.

## Motivation

Self-introspection debug pages or zPages are in-process web pages that display collected data from the process they are attached to. They are used to provide in-process diagnostics without the need of any backend to examine traces or metrics. Various implementations of zPages are widely used in many environments. The standard extensible implementation of zPages in OpenTelemetry will benefit everybody.

## Explanation

This OTEP is a request to get a general approval for zPages development as an experimental feature [open-telemetry/opentelemetry-specification#62](https://github.com/open-telemetry/opentelemetry-specification/pull/632). See [opencensus.io/zpages](https://opencensus.io/zpages/) for the overview of zPages.

## Internal details

Implementation of zPages includes multiple components - data collection (sampling, filtering, configuration), storage and aggregation, and a framework to expose this data.

This is a request for a general direction approval. There are a few principles for the development:

> 1. zPages MUST NOT be hardcoded into OpenTelemetry SDK.

Reviewer 3: What do you mean by "not be hardcoded"? Something like this?
```suggestion
1. zPages MUST NOT be enabled or configured by default in the OpenTelemetry SDK.
```

Reviewer 3: Or does this mean it must be supplied in a separate package/code unit rather than "weaved into" the SDK?

Reviewer 3: @SergeyKanzhelev Please clarify.

Reviewer 4: I agree with your interpretation about making zPages a separate package / code unit, at the level of a span process (for tracing). In OpenCensus, the zPages support was baked-in to the default SDK tracer. This relates to https://github.com/open-telemetry/opentelemetry-specification/issues/373 about having access to live, unfinished spans.

> 2. OpenTelemetry implementation of zPages MUST be split as two separate components - one for data, another for rendering. So that, for example, data providers could be also integrated into other rendering frameworks.

Reviewer 3: ```suggestion
2. OpenTelemetry implementation of zPages MUST be split into two separate components - one for data collection and handling, another for rendering. This way, for example, data providers could be also integrated into other rendering frameworks.
```
